api/v1alpha1/raycluster_types.go

```diff
@@ -1,31 +1,83 @@
 package v1alpha1
 
 import (
+	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
-// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
-
-// RayClusterSpec defines the desired state of RayCluster
+// RayClusterSpec defines the desired state of a RayCluster resource.
 type RayClusterSpec struct {
-	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
-	// Important: Run "make" to regenerate code after modifying this file
+	// Image used to launch head and worker nodes.
+	// +kubebuilder:default={repository: "ray-project/ray-ml", tag: "1.2.0-cpu"}
+	Image OCIImageDefinition `json:"image"`
+
+	// WorkerReplicaCount configures the total number of workers in the cluster.
+	// +kubebuilder:default=2
+	WorkerReplicaCount int `json:"workerReplicaCount"`
+
+	// HeadPort is the port of the head ray process.
+	// +kubebuilder:default=6379
+	HeadPort int `json:"port"`
+
+	// RedisShardPorts is a list of ports for non-primary Redis shards.
+	// +kubebuilder:default={6380,6381}
+	RedisShardPorts []int `json:"redisShardPorts"`
+
+	// ObjectManagerPort is the raylet port for the object manager.
+	// +kubebuilder:default=2384
+	ObjectManagerPort int `json:"objectManagerPort"`
+
+	// NodeManagerPort is the raylet port for the node manager.
+	// +kubebuilder:default=2385
+	NodeManagerPort int `json:"nodeManagerPort"`
+
```

Comment:
I don't understand the question, would you please elaborate.

Reply:
Ray will call frameworks which communicate out of band with Ray. They will introduce other ports that may need to be opened. I think, compared to Spark, this may be especially important for Ray if we plan to lock down network communication.

Reply:
It just sounds like the network policies have to accommodate for those extra ports. I look to Torch as an example and see what that presents.